src/KioskMode/ControlBuilder.php

```diff
@@ -0,0 +1,81 @@
+<?php
+/* Copyright (c) 2018 - Richard Klees <richard.klees@concepts-and-training.de> - Extended GPL, see LICENSE */
+
+namespace ILIAS\KioskMode;
+
+use ILIAS\UI;
+
+/**
+ * Build controls for the view.
+ */
+interface ControlBuilder {
+	/**
+	 * An exit control allows the user to gracefully leave the object providing
+	 * the kiosk mode.
+	 *
+	 * @throws \LogicException if view wants to introduce a second exit button.
+	 */
+	public function exit(string $command) : ControlBuilder;
+
+	/**
+	 * A next control allows the user to progress to the next item in the object.
+	 *
+	 * @throws \LogicException if view wants to introduce a second next button.
+	 */
+	public function next(string $command, int $parameter) : ControlBuilder;
+
+	/**
+	 * A previous control allows the user to go back to the previous item in the object.
+	 *
+	 * @throws \LogicException if view wants to introduce a second previous button.
+	 */
+	public function previous(string $command, int $parameter) : ControlBuilder;
+
+	/**
+	 * A done control allows the user to mark the object as done.
+	 *
+	 * @throws \LogicException if view wants to introduce a second previous button.
+	 */
+	public function done(string $command, int $parameter) : ControlBuilder;
+
+	/**
+	 * A generic control needs to have a label that tells what it does.
+	 */
+	public function generic(string $label, string $command, int $parameter) : ControlBuilder;
+
+	/**
+	 * A toggle can be used to switch some behaviour in the view on or of.
+	 */
+	public function toggle(string $label, string $on_command, string $off_command) : ControlBuilder;
+
+	/**
+	 * A mode control can be used to switch between different modes in the view.
+	 *
+	 * Uses the indizes of the labels in the array as parameter for the command.
+	 */
+	public function mode(string $command, array $labels) : ControlBuilder;
+
+	/**
+	 * A locator allows the user to see the path leading to her current location and
+	 * jump back to previous items on that path.
+	 *
+	 * The command will be enhanced with a parameter defined in the locator builder.
+	 *
+	 * @throws \LogicException if view wants to introduce a second locator.
+	 */
+	public function locator(string $command) : LocatorBuilder;
```

Comment:
I think this is missing a explanation about how these controls relate to the UI framework. We have a Breadcrumb element there, which fits the same needs. General non kiosk mode views use Breadcrumbs and here I have to duplicate the same logic using a locator. This is the duplication of all interactive elements in components views and this will lead to high maintenance overhead imo.

Reply:
The relation is this: A player could decide to render the abstract control demanded by the `View` via a Breadcrumb from the UI-framework. Most possibly, the service will provide an implementation of ControlBuilder, that will do just that. Point is: the user gets to decide how the control is in fact rendered.